Packages/com.unity.inputsystem/InputSystem/Devices/ScreenKeyboard.cs

```diff
@@ -0,0 +1,290 @@
+    ///      this matches the behavior in old TouchScreenKeyboard implementation)
+    ///   - [iOS] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    ///   - [WSA] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    /// </summary>
+    internal abstract class ScreenKeyboard : IDisposable
+    {
+        protected ScreenKeyboardState m_KeyboardState;
+        protected ScreenKeyboardShowParams m_ShowParams;
+        private ScreenKeyboardCallbacks m_Callbacks;
+
+        /// <summary>
+        /// Returns the state of the keyboard
+        /// </summary>
+        public ScreenKeyboardState state
+        {
+            get => m_KeyboardState;
+        }
+
+        /// <summary>
+        /// Cleanup any platform specific resources
+        /// </summary>
+        public virtual void Dispose()
+        {
+        }
+
+        /// <summary>
+        /// Show the screen keyboard with customized options.
+        /// </summary>
+        /// <param name="showParams"></param>
+        /// <param name="callbacks"></param>
+        public void Show(ScreenKeyboardShowParams showParams = default, ScreenKeyboardCallbacks callbacks = default)
+        {
+            m_ShowParams = showParams;
+            m_Callbacks = callbacks;
+            if (m_ShowParams.initialText == null)
+                m_ShowParams.initialText = String.Empty;
+            if (m_ShowParams.placeholderText == null)
+                m_ShowParams.placeholderText = String.Empty;
+            InternalShow();
+        }
+
+        /// <summary>
+        /// Hide screen keyboard
+        /// </summary>
+        public void Hide()
+        {
+            InternalHide();
+        }
+
+        protected abstract void InternalShow();
+
+        protected abstract void InternalHide();
+
+        protected void ReportInputFieldChange(string text)
+        {
+            m_Callbacks.inputFieldTextChanged?.Invoke(text);
+        }
+
+        protected void ReportStateChange(ScreenKeyboardState state)
+        {
+            if (state != m_KeyboardState)
+            {
+                m_KeyboardState = state;
+                m_Callbacks.stateChanged?.Invoke(state);
+            }
+        }
+
+        protected void ReportSelectionChange(int start, int length)
+        {
+            if (m_KeyboardState != ScreenKeyboardState.Visible)
+                return;
+
+            var selection = new RangeInt(start, length);
+            m_Callbacks.inputFieldSelectionChanged?.Invoke(selection);
+        }
+
+        /// <summary>
+        /// Modifies text in screen keyboard's input field.
+        /// You can use this propery even if input field is hidden
+        /// </summary>
+        public abstract string inputFieldText { set; get; }
```

Comment:
Another side note from what I noticed in current TouchScreenKeyboard: what happens to selection when you set text?

Reply:
As per our discussion, the most logical thing to do, is set selection at the end of the text